plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/StaticSelector.java

```diff
@@ -137,20 +125,90 @@ private static void addNamedGroups(Pattern pattern, HashSet<String> variables)
         }
     }
 
-    private void addVariableValues(Pattern pattern, String candidate, Map<String, String> mapping)
+    private interface SelectionMatcher
+    {
+        boolean matches(SelectionCriteria criteria);
+
+        void populateVariables(SelectionCriteria criteria, Map<String, String> variables);
+    }
+
+    private record BasicMatcher(Predicate<SelectionCriteria> matchFn)
+            implements SelectionMatcher
+    {
+        @Override
+        public boolean matches(SelectionCriteria criteria)
+        {
+            return matchFn.test(criteria);
+        }
+
+        @Override
+        public void populateVariables(SelectionCriteria criteria, Map<String, String> variables)
+        {
+            // no-op
+        }
+    }
+
+    private static class PatternMatcher
+            implements SelectionMatcher
     {
-        for (String key : variableNames) {
-            Matcher keyMatcher = pattern.matcher(candidate);
-            if (keyMatcher.find()) {
-                try {
-                    String value = keyMatcher.group(key);
+        private final Set<String> variableNames;
+        private final Pattern pattern;
+        private final Function<SelectionCriteria, String> valueExtractor;
+
+        PatternMatcher(Set<String> variableNames, Pattern pattern, Function<SelectionCriteria, String> valueExtractor)
```

Comment:
In this project, I generally observe a pattern where methods from private classes are public. The idea, I believe, is that they constitute the public API of a class that happens to be private.

Reply:
Sure, added public. Given the limited scope of the class, `package-private` vs `public` is functionally identical so I don't have much opinion either way.